docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -569,80 +294,279 @@ helm template --values /tmp/values.yaml .
-    // required by the user. If the list has more than one item `IsList` must be true. (optional)
-    DefaultValues []string `json:"defaultValues"`
-}
+A parameter definition is an object with following schema:
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: ConfigManagementPlugin
+metadata:
+  name: helm
+spec:
+  parameters:
+    dynamic:
+      command: 
+      - sh
+      - -c
+      - |
+        # Use yq to generate a list of parameters. Then use jq to convert that list of parameters to a parameters
+        # announcement list.
+        yq e -o=p values.yaml | jq -nR '
+          [{
+            name: "helm-parameters",
+            title: "Helm Parameters",
+            tooltip: "Parameters to override when generating manifests with Helm",
+            collectionType: "map",
+            map: (inputs | capture("(?<key>.*) = (?<value>.*)") | from_entries)
+          }]'
 ```
 
-```go
-type ParametersAnnouncement []ParameterDefinition
+For a Helm chart with only an `image.repository` and `image.tag` in values.yaml, the parameter announcement would look
+like this:
+
+```json
+[
+  {
+    "name": "helm-parameters",
+    "collectionType": "map",
+    "title": "Helm Parameters",
+    "tooltip": "Parameters to override when generating manifests with Helm",
+    "map": {
+      "image.repository": "my.company.com/gcr-proxy/heptio-images/ks-guestbook-demo",
+      "image.tag": "0.1"
+    }
+  }
+]
 ```
 
-Example:
+Before sending a parameters announcement to the UI, the CMP server will combine the static and dynamic parameters.
+(Behind the scenes, the list is actually communicated to the UI via gRPC, but they're presented here as JSON for 
+readability.)
 
 ```json
+
 [
   {
     "name": "values-files",
-    "type": "enum",
-    "uiConfig": "{\"values\": [\"values.yaml\"]}"
+    "title": "Values Files",
+    "collectionType": "array"
   },
   {
-    "name": "image",
-    "type": "image",
-    "section": "Helm Parameters"
+    "name": "helm-parameters",
+    "collectionType": "map",
+    "title": "Helm Parameters",
+    "tooltip": "Parameters to override when generating manifests with Helm",
+    "map": {
+      "image.repository": "my.company.com/gcr-proxy/heptio-images/ks-guestbook-demo",
+      "image.tag": "0.1"
+    }
   }
 ]
 ```
 
+This is the full parameters announcement schema as Go types.
+
+```go
+package cmp
+
+// ParameterItemType is the primitive data type of each of the parameter's value (or each of its values, if it's an array or
+// a map).
+type ParameterItemType string
+
+const (
+	ParameterTypeString  ParameterItemType = "string"
+	ParameterTypeNumber  ParameterItemType = "number"
+	ParameterTypeBoolean ParameterItemType = "boolean"
+)
+
+// ParameterCollectionType is a parameter's value's type - a single value (like a string) or a collection (like an array or a
+// map).
+type ParameterCollectionType string
+
+const (
+	ParameterValueTypeString ParameterCollectionType = "string"
```

Comment:
maybe we could exclude the `ParameterValueTypeString` as it is confusing with `ParameterTypeString`. I guess it is better to just have the `ParameterValueTypeMap`, `ParameterValueTypeArray` collection types if if omitted it is considered as string

Reply:
I'm good with that. It means there will be areas on the code like this

```go
if announcement.CollectionType == "" {
  // do things with announcement.string
}
```

Which might not read quite as nicely as this:

```go
if announcement.CollectionType == ParameterValueTypeString {
  // do things with announcement.string
}
```

But I think it's fine either way.

Reply:
Updated!